test: add zone.js test to saucelabs

JiaLiPassion opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [x] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


There are several tests of Zone.js run inside saucelabs when zone.js were not merged into angular repo, those tests are not put into the angular circle ci, in this PR, I added them back.

devversion reviewed (changes requested):
I'm unsure if it is right to add the zone.js tests to the legacy saucelabs tests. Instead, I'd expect them at least to be part of the Saucelabs Bazel test suite.

Also @gregmagolan and @josephperrott are working on improving the overall Saucelabs test experience with Bazel. So it might be good to coordinate with them. I don't think their work landed yet.

JiaLiPassion commented:
@devversion, thank you for the information, I will check with @gregmagolan and @josephperrott to get help.

josephperrott commented:
@JiaLiPassion I think it makes sense to address this once #33927 lands.  We should be running these tests via bazel.

JiaLiPassion commented:
@josephperrott , got it, thanks, I will keep checking the status of #33927.

AndrewKushnir commented:
Hi @JiaLiPassion, is this PR still relevant?

gkalpak commented on .circleci/config.yml:
```diff
@@ -722,6 +722,22 @@ jobs:
       - run: yarn karma start ./karma-js.conf.js --single-run --browsers=${KARMA_JS_BROWSERS}
+        # run zone.js saucelabs test
+      - run: cd packages/zone.js
+      - run: yarn
```
Shouldn't we be using the same command as [test_zone.js][1]?

[1]: https://github.com/angular/angular/blob/cd927a27e158e10f5b407dc19337cbff7f8d34bb/.circleci/config.yml#L827

JiaLiPassion replied:
Yeah, I am using `yarn --cwd` now. thanks.

gkalpak commented on .circleci/config.yml:
```diff
@@ -722,6 +722,22 @@ jobs:
+      - run: yarn karma start karma-dist-sauce-jasmine3.conf.js --single-run --error-policy=lazy
+      - run: node simple-server.js 2>&1> server.log&
+      - run: node ./test/webdriver/test.sauce.js
+      - run: yarn add jasmine@3.0.0 jasmine-core@3.0.0 mocha@5.0.1
```
Why do we have to install these "manually"? Couldn't they be part of `package.json`?

JiaLiPassion replied:
Sure, I have added them to the package.json, thanks.

gkalpak commented on .circleci/config.yml:
```diff
@@ -722,6 +722,22 @@ jobs:
+      - run: yarn add jasmine@3.0.0 jasmine-core@3.0.0 mocha@5.0.1
+      - run: yarn karma start karma-dist-sauce-jasmine3.conf.js --single-run
+      - run: yarn karma start karma-build-sauce-selenium3-mocha.conf.js --single-run
+      - run: yarn karma start  karma-evergreen-dist-sauce-jasmine.conf.js --single-run
```
```suggestion
      - run: yarn karma start karma-evergreen-dist-sauce-jasmine.conf.js --single-run
```
Also, since there are too many commands to test `zone.js` on soucelabs, could we move them into a script inside `packages/zone.js/`? It makes the CI config easier to follow.

JiaLiPassion replied:
So now since only `CI_REQUIRED` browsers need to be tested, I only left one test here.

JiaLiPassion committed:
test(zone.js): add zone.js to saucelab test

When migrating zone.js into the angular repo, the saucelab legacy test is disabled,
and in this PR, the zone.js saucelab legacy browser tests are added back and use
the same configuration with the angular.

This PR also disable some cases in the `IE/Edge/Android` platform since
these cases cause flaky results.

JiaLiPassion commented:
@gkalpak , Thanks for the review, I have updated the PR, could you check it again?

gkalpak reviewed (changes requested):
Sorry for my late response. I completely missed your previous message, @JiaLiPassion 😇
I left a bunch of minor suggestions and questions, but this mostly looks good.

jessicajaniuk commented:
@JiaLiPassion Any updates on this PR?

atscott commented:
Closing as this has gotten out of date. @JiaLiPassion If this is still something you want to do, please open a new PR.

JiaLiPassion commented:
@atscott , sorry I forgot about this one, I would like to reopen it since there are some discussions/review comments here, so it is easy to track.

JiaLiPassion commented on packages/zone.js/karma-build.conf.js:
```diff
@@ -8,11 +8,12 @@
 module.exports = function(config) {
   require('./karma-base.conf.js')(config);
-  config.files.push('node_modules/core-js-bundle/index.js');
+  const defaultRule = config.files.splice(-1, 1);
```
The reason is we need to load these `env-setup/polyfills` file before import `zone.js`, otherwise some test related `environment vars` will not be imported.

JiaLiPassion commented on packages/zone.js/sauce.conf.js:
```diff
@@ -1,136 +1,34 @@
-
-    plugins: ['karma-*']
+    transports: ['websocket', 'polling'],
+    // plugins: ['karma-*']
```
Removed, thank you.

JiaLiPassion commented on packages/zone.js/test/browser-zone-saucelab-setup.ts:
```diff
@@ -0,0 +1,35 @@
+  (window as any)['__Zone_enable_cross_context_check'] = true;
+  (window as any)[zoneSymbol('fakeAsyncAutoFakeAsyncWhenClockPatched')] = true;
+  (window as any).isSaucelab = true;
+  // if (typeof Promise === 'undefined') {
+  //   require('core-js-bundle');
+  // }
```
Removed, thank you.

JiaLiPassion commented on packages/zone.js/karma-build.conf.js:
```diff
@@ -8,11 +8,12 @@
   require('./karma-base.conf.js')(config);
-  config.files.push('node_modules/core-js-bundle/index.js');
+  const defaultRule = config.files.splice(-1, 1);
+  // config.files.push('node_modules/core-js-bundle/index.js');
```
Removed, thank you.

JiaLiPassion commented on packages/zone.js/package.json:
```diff
@@ -15,11 +15,31 @@
+    "systemjs": "^0.19.37",
+    "typescript": "4.1.2",
+    "webdriver-manager": "^12.0.6",
+    "webdriverio": "^4.8.0",
```
To run the legacy karma tests in the saucelab env, now I am using `karma` runner directly instead of using the `bazel` rule, the zone.js legacy karma test need those libs to prepare the environment and the test.

JiaLiPassion commented on packages/zone.js/karma-base.conf.js:
```diff
@@ -13,20 +13,20 @@ module.exports = function(config) {
     files: [
       'node_modules/systemjs/dist/system-polyfills.js', 'node_modules/systemjs/dist/system.src.js',
       'node_modules/whatwg-fetch/fetch.js',
```
this is the legacy `zone.js` settings.I will clean up the `karma config` in the other PR,

JiaLiPassion commented on packages/zone.js/.gitignore:
```diff
@@ -1 +1,2 @@
 build
+dist/*.js
```
`zone.js:npm_package` rule will generate `js` bundles under `dist/zone.js-dist/zone.js/dist` such as   `dist/zone.js-dist/zone.js/dist/zone.js`, so no file will be generated under `packages/zone.js/dist` any longer if we are using everything with bazel, but if we test legacy browser with saucelab locally, we need to generate zone.js bundles under `packages/zone.js/dist` in the current setting, so this `dist/*.js` in the `.gitignore` just to prevent the developer add those files into git.

JiaLiPassion commented on .circleci/config.yml:
```diff
@@ -668,6 +672,7 @@ jobs:
           command: |
             browsers=$(node -e 'console.log(require("./browser-providers.conf").sauceAliases.CI_REQUIRED.join(","))')
             yarn karma start ./karma-js.conf.js --single-run --browsers=${browsers}
+            yarn karma start ./packages/zone.js/karma-dist-sauce-jasmine.conf.js --single-run
```
Yes, you are right, I added the browsers here, thank you.

JiaLiPassion commented on .circleci/config.yml:
```diff
@@ -647,6 +647,7 @@ jobs:
       - run: yarn tsc -p packages --module UMD
       - run: yarn tsc -p modules --module UMD
       - run: yarn bazel build //packages/zone.js:npm_package
+      - run: yarn --cwd=./packages/zone.js install && yarn --cwd=./packages/zone.js tsc -p . --module UMD
```
Sure, I update the `install` and also separate the command into `2 run steps`.

jessicajaniuk commented:
@devversion @gkalpak Can you take another look at this PR so it gets unblocked? @JiaLiPassion Can you also rebase to fix the conflicts?

JiaLiPassion commented:
@jessicajaniuk , sure, I just rebased this PR.

devversion commented on packages/zone.js/package.json:
```diff
@@ -19,7 +19,27 @@
+    "nodejs-websocket": "^1.2.0",
+    "rxjs": "^6.2.1",
+    "selenium-webdriver": "^3.4.0",
+    "systemjs": "^0.19.37",
```
I'm personally a little afraid that we start leveraging this nested `package.json` even more, making it look like it's well-integrated into the workspace (like with Yarn workspace) -- it's not!

ZoneJS is part of the monorepo and relies on the top-level/main node modules for building, testing etc, but then we have this legacy construct here. I think we should not start building on-top of it even more? WDYT?

Also there is work in progress for leveraging Bazel more for Saucelabs tests. This adds to the legacy job right? so maybe not worth at all right now?

gregmagolan replied:
Better saucelabs tests under Bazel is planned for this half but not in progress yet. We're working on the angular/aio & angular-cli repos atm.

devversion reviewed (changes requested):
I added a comment on why I think it might not be worth doing right now, but regardless, thanks @JiaLiPassion for your work on this and all the other ZoneJS PRs 😄 

<sup>I feel like cleaning up the legacy setup might be a super valuable task</sup>
